language/move-lang/src/unit_test/filter_test_members.rs

```diff
@@ -0,0 +1,170 @@
+
+use crate::{
+    parser::ast as P,
+    shared::{known_attributes, CompilationEnv},
+    unit_test::Context,
+};
+use move_ir_types::location::Loc;
+
+//***************************************************************************
+// Filtering of test-annotated module members
+//***************************************************************************
+
+// This filters out all test, and test-only annotated module member from `prog` if the `test` flag
+// in `compilation_env` is not set. If the test flag is set, no filtering is performed, and instead
+// a test plan is created for use by the testing framework.
+pub fn program(compilation_env: &mut CompilationEnv, prog: P::Program) -> P::Program {
+    let mut context = Context::new(compilation_env);
+
+    let P::Program {
+        source_definitions,
+        lib_definitions,
+    } = prog;
+
+    let lib_definitions: Vec<_> = lib_definitions
+        .into_iter()
+        .filter_map(|def| filter_tests_from_definition(&mut context, def))
+        .collect();
+
+    let source_definitions: Vec<_> = source_definitions
+        .into_iter()
+        .filter_map(|def| filter_tests_from_definition(&mut context, def))
+        .collect();
+
+    P::Program {
+        source_definitions,
+        lib_definitions,
+    }
+}
+
+fn filter_tests_from_definition(
+    context: &mut Context,
+    def: P::Definition,
+) -> Option<P::Definition> {
+    match def {
+        P::Definition::Module(m) => filter_tests_from_module(context, m).map(P::Definition::Module),
+        P::Definition::Address(attributes, loc, addr, ms) => {
+            let ms: Vec<_> = ms
+                .into_iter()
+                .filter_map(|m| filter_tests_from_module(context, m))
+                .collect();
+            if !ms.is_empty() && !should_remove_node(context.env, &attributes) {
+                Some(P::Definition::Address(attributes, loc, addr, ms))
+            } else {
+                None
+            }
+        }
+        P::Definition::Script(P::Script {
+            attributes,
+            uses,
+            constants,
+            function,
+            specs,
+            loc,
+        }) => {
+            let script_attributes = attributes
+                .iter()
+                .chain(uses.iter().flat_map(|use_decl| &use_decl.attributes))
+                .chain(function.attributes.iter())
+                .chain(specs.iter().flat_map(|spec| &spec.value.attributes));
+
+            let errors: Vec<_> = script_attributes
+                .map(|attr| {
+                    test_attributes(attr)
+                        .into_iter()
+                        .map(|(loc, _)| {
+                            let msg = "Testing attributes are not allowed in scripts.";
+                            (loc, msg.to_string())
+                        })
+                        .collect::<Vec<_>>()
+                })
+                .filter(|x| !x.is_empty())
+                .collect();
```

Comment:
This still feels really weird. I think you just want to loop over each attribute individually, and error if you see one. Each one should get its own error.

Reply:
Updated them so each one gets its own error. I'm not a big fan of looping over them here and instead building the errors up this way, but I can change it over to that if you would prefer that after this change.

Reply:
It's all the same really. The flatten is fine :)